config/management/operational/src/validator_config.rs

```diff
@@ -1,173 +1,144 @@
+            host: self.host,
+            validator_config: self.validator_config,
+            validator_address: None,
+            fullnode_address: None,
+        };
 
-        // Rotate key in storage and network address
-        let new_validator_network_key = storage.rotate_key(VALIDATOR_NETWORK_KEY)?;
-        let new_validator_network_key = StorageWrapper::x25519(new_validator_network_key)?;
-        validator_address.rotate_noise_public_key(
-            &validator_config.validator_network_identity_public_key,
-            &new_validator_network_key,
-        );
-        let raw_enc_validator_address = encode_validator_address(
-            validator_address,
-            &self.owner_account,
-            sequence_number,
-            &TEST_SHARED_VAL_NETADDR_KEY,
-            TEST_SHARED_VAL_NETADDR_KEY_VERSION,
-            idx_addr,
-        )?;
-        validator_config.validator_network_identity_public_key = new_validator_network_key;
-        validator_config.validator_network_address = raw_enc_validator_address;
+        set_validator_config.execute().map(|txn_ctx| (txn_ctx, key))
+    }
+}
 
-        // Submit the transaction with a reconfigure
-        let transaction_context = update_config_and_reconfigure(
-            storage_name,
-            &self.validator_backend.validator_backend,
-            &client,
-            self.owner_account,
-            self.chain_id,
-            sequence_number,
-            validator_config,
-        )?;
+fn decode_validator_address(
+    address: RawEncNetworkAddress,
+    account: &AccountAddress,
+    key: &Key,
+    addr_idx: u32,
+) -> Result<NetworkAddress, Error> {
+    let enc_addr = EncNetworkAddress::try_from(&address).map_err(|e| {
+        Error::UnexpectedError(format!(
+            "Failed to decode network address {}",
+            e.to_string()
+        ))
+    })?;
+    let raw_addr = enc_addr.decrypt(key, account, addr_idx).map_err(|e| {
+        Error::UnexpectedError(format!(
+            "Failed to decrypt network address {}",
+            e.to_string()
+        ))
+    })?;
+    decode_address(raw_addr)
+}
 
-        Ok((transaction_context, new_validator_network_key))
-    }
+fn decode_address(raw_address: RawNetworkAddress) -> Result<NetworkAddress, Error> {
+    let network_address = NetworkAddress::try_from(&raw_address).map_err(|e| {
+        Error::UnexpectedError(format!(
+            "Failed to decode network address {}",
+            e.to_string()
+        ))
+    })?;
+    let protocols = network_address
+        .as_slice()
+        .iter()
+        .filter(|protocol| match protocol {
+            Protocol::Dns(_)
+            | Protocol::Dns4(_)
+            | Protocol::Dns6(_)
+            | Protocol::Ip4(_)
+            | Protocol::Ip6(_)
+            | Protocol::Memory(_)
+            | Protocol::Tcp(_) => true,
+            _ => false,
+        })
+        .cloned()
+        .collect();
+    Ok(NetworkAddress::new(protocols))
```

Comment:
I kinda feel like all of this code belongs under network, but that feels like another PR so we don't end up modifying too many crates here. Since we seem to agree, I'll follow up in another commit / PR.